cpp/src/arrow/array/array_test.cc

```diff
@@ -80,6 +80,22 @@ class TestArray : public ::testing::Test {
   MemoryPool* pool_;
 };
 
+void CheckDictionaryNullCount(const std::shared_ptr<DataType>& dict_type,
+                              const std::string& input_dictionary_json,
+                              const std::string& input_index_json,
+                              const int64_t& expected_null_count,
+                              const int64_t& expected_logical_null_count,
+                              bool expected_may_have_nulls,
+                              bool expected_may_have_logical_nulls) {
+  std::shared_ptr<arrow::Array> arr =
+      DictArrayFromJSON(dict_type, input_index_json, input_dictionary_json);
+
+  ASSERT_EQ(expected_null_count, arr->null_count());
+  ASSERT_EQ(expected_logical_null_count, arr->ComputeLogicalNullCount());
+  ASSERT_EQ(expected_may_have_nulls, arr->data()->MayHaveNulls());
+  ASSERT_EQ(expected_may_have_logical_nulls, arr->data()->MayHaveLogicalNulls());
```

Comment:
Nit: prefer expectations as RHS
```suggestion
  ASSERT_EQ(arr->null_count(), expected_null_count);
  ASSERT_EQ(arr->ComputeLogicalNullCount(), expected_logical_null_count);
  ASSERT_EQ(arr->data()->MayHaveNulls(), expected_may_have_nulls);
  ASSERT_EQ(arr->data()->MayHaveLogicalNulls(), expected_may_have_logical_nulls);
```